src/components/SeatContest.tsx

```diff
@@ -125,101 +132,110 @@ class SeatContest extends React.Component<Props, State> {
+  public removeCandidateFromVote = (id: string) => {
     const { contest, vote } = this.props
-    const { value } = event.target as HTMLInputElement
-    const targetIsSelected = vote && value === vote.id
-    this.selectCandidate(
-      targetIsSelected
-        ? undefined
-        : contest.candidates.find(candidate => candidate.id === value)
-    )
+    const newVote = vote.filter(c => c.id !== id)
+    this.props.updateVote(contest.id, newVote)
   }
 
-  public handleChangeVoteAlert = (attemptedVoteCandidateName: string) => {
-    this.setState({ attemptedVoteCandidateName })
+  public handleUpdateSelection = (event: InputEvent) => {
+    const { vote } = this.props
+    const id = (event.target as HTMLInputElement).value
+    const candidate = this.findCandidateById(vote, id)
+    if (!!candidate) {
+      if (candidate.isWriteIn) {
+        this.setState({ candidatePendingRemoval: candidate })
+      } else {
+        this.removeCandidateFromVote(id)
+      }
+    } else {
+      this.addCandidateToVote(id)
+    }
+  }
+
+  public handleChangeVoteAlert = (
+    attemptedOverVoteCandidate: OptionalCandidate
+  ) => {
+    this.setState({ attemptedOverVoteCandidate })
   }
 
   public closeAttemptedVoteAlert = () => {
-    this.setState({ attemptedVoteCandidateName: '' })
+    this.setState({ attemptedOverVoteCandidate: undefined })
   }
 
-  public handleSelectWriteInCandidate = () => {
-    const { vote } = this.props
-    if (!!vote && vote.name === this.state.writeInCandidateName) {
-      this.selectCandidate(undefined)
-    } else {
-      this.toggleWriteInCandidateModal(true)
-    }
+  public confirmRemovePendingWriteInCandidate = () => {
+    this.removeCandidateFromVote(this.state.candidatePendingRemoval!.id)
+    this.clearCandidateIdPendingRemoval()
   }
-  public handleWriteInCandidateDisabledClick = () => {
-    const { vote } = this.props
-    const { writeInCandidateName } = this.state
-    if (!!vote && vote.name !== writeInCandidateName) {
-      this.handleChangeVoteAlert(writeInCandidateName || 'a write-in candidate')
-    }
+
+  public clearCandidateIdPendingRemoval = () => {
+    this.setState({ candidatePendingRemoval: undefined })
+  }
+
+  public initWriteInCandidate = () => {
+    this.toggleWriteInCandidateModal(true)
   }
 
   public normalizeName = (name: string) =>
     name
       .trim()
       .replace(/\t+/g, ' ')
       .replace(/\s+/g, ' ')
-  public selectWriteInCandidate = () => {
-    const writeInCandidateName = this.normalizeName(
+
+  public addWriteInCandidate = () => {
+    const { contest, vote } = this.props
+    const normalizedCandidateName = this.normalizeName(
       this.state.writeInCandidateName
     )
-    this.setState({ writeInCandidateName })
-    this.selectCandidate({
-      id: 'writeInCandidate',
-      name: writeInCandidateName,
-    })
+    this.props.updateVote(contest.id, [
+      ...vote,
+      {
+        id: camelCase(normalizedCandidateName),
```

Comment:
I think the `id` here should probably be `write-in-1`, `write-in-2`, etc. I can imagine collisions between write-in names and existing candidates, and in general I think it may be a potential problem to use IDs that are decided by user input.

Reply:
Good call. Fixed by prefixing camelCased candidate name with `write-in__`.